src/java_bytecode/expr2java.h

```diff
@@ -50,4 +53,52 @@ class expr2javat:public expr2ct
 std::string expr2java(const exprt &expr, const namespacet &ns);
 std::string type2java(const typet &type, const namespacet &ns);
 
+/// Convert floating point number to a string without printing
+/// unnecessary zeros. Prints decimal if precision is not lost.
+/// Prints hexadecimal otherwise, and/or java-friendly NaN and Infinity
+template <typename float_type>
+std::string floating_point_to_java_string(float_type value)
+{
+  const auto is_float = std::is_same<float_type, float>::value;
+  static const std::string class_name = is_float ? "Float" : "Double";
+  if(std::isnan(value))
+    return class_name + ".NaN";
+  if(std::isinf(value) && value >= 0.)
+    return class_name + ".POSITIVE_INFINITY";
+  if(std::isinf(value) && value <= 0.)
+    return class_name + ".NEGATIVE_INFINITY";
+  // Using ostringstream to get string without trailing zeros
+  const std::string decimal = [&] { // NOLINT
+    std::ostringstream raw_stream;
+    raw_stream << value;
+    const auto raw_decimal = raw_stream.str();
+    if(raw_decimal.find('.') == std::string::npos)
+      return raw_decimal + ".0";
+    return raw_decimal;
+  }();
+  const bool is_lossless = [&] { // NOLINT
+    if(value == std::numeric_limits<float_type>::min())
+      return true;
+    try
+    {
+      return std::stod(decimal) == value;
+    }
+    catch(std::out_of_range)
+    {
+      return false;
+    }
+  }();
+  const std::string lossless = [&]() -> std::string { // NOLINT
```

Comment:
I don't think this is a very clear variable name, perhaps `lossless_representation`